### PR TITLE
support $LATEST placeholder in BigQueryType, fix #474

### LINKED ISSUE #474: Use latest data/partition in type-safe BigQuery annotations
`@BigQueryType.fromTable` & `@BigQueryType.fromQuery` currently requires a specific data/partition, i.e. `@BigQueryType.fromQuery("SELECT * FROM `log_%s`", "20170301")` but it'd be nice to use the latest tables available.

We could have a magic string / constant to indicate an order by table date wildcard. So:
- `val LATEST = "[LATEST]"
- `fromTable("dataset.prefix_%s", LATEST) would match newest table in `dataset.prefix_*`

However, join case could be trickier
- `fromQuery("SELECT * FROM dataset1.prefix1_%s JOIN dataset2.prefix2_%s", LATEST, LATEST)` may join tables from different dates depending on availability.
- unless if we have another keyword that indicates matching only the latest available in all sources, like "LATEST_ALL"?

### PR DESCRIPTION
**Pull request by nevillelyh**

### CODE REVIEW

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryPartitionUtil.scala [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    }
+    val m2 = QUERY_TABLE_SPEC_STANDARD.matcher(sqlQuery)
+    while (m2.find()) {
+      val t = m2.group(0)
+      b += (s"`$t`" -> BigQueryIO.parseTableSpec(t.replaceFirst("\\.", ":")))
+    }
+    b.result()
+  }
+
+  private def getPartitions(bq: BigQueryClient, tableRef: TableReference): Set[String] = {
```
Comment (marklaw000) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryPartitionUtil.scala]: isn't partition in bigquery is another dimension of the table. maybe getTableIds will be a nicer name?

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryPartitionUtil.scala [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    if (ref.getTableId.endsWith("$LATEST")) {
+      val partitions = getPartitions(bq, ref)
+      require(partitions.nonEmpty, s"Cannot find latest partition for $tableSpec")
+      tableSpec.replace("$LATEST", partitions.max)
+    } else {
+      tableSpec
+    }
+  }
+
+  def main(args: Array[String]): Unit = {
```
Comment (ravwojdyla) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryPartitionUtil.scala]: ?

Comment (nevillelyh) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryPartitionUtil.scala]: LOL

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryClient.scala [hunk 3]
```diff
@@ -225,6 +216,21 @@ class BigQueryClient private (private val projectId: String,
     bigquery.tables().get(p, table.getDatasetId, table.getTableId).execute()
   }
 
+  /** Get list of tables in a dataset. */
+  def getTables(projectId: String, datasetId: String): Seq[TableReference] = {
+    val b = Seq.newBuilder[TableReference]
+    val req = bigquery.tables().list(projectId, datasetId)
+    var rep = req.execute()
```
Comment (ravwojdyla) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryClient.scala]: `ScioUtil.executeWithBackOff(req)`

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryClient.scala [hunk 4]
```diff
@@ -225,6 +216,21 @@ class BigQueryClient private (private val projectId: String,
...
+    rep.getTables.asScala.foreach(t => b += t.getTableReference)
+    while (rep.getNextPageToken != null) {
+      rep = req.setPageToken(rep.getNextPageToken).execute()
```
Comment (ravwojdyla) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryClient.scala]: + `executeWithBackOff`

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryPartitionUtil.scala [hunk 5]
```diff
@@ -0,0 +1,76 @@
+package com.spotify.scio.bigquery
```
Comment (ravwojdyla) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/BigQueryPartitionUtil.scala]: Header missing.

File: scio-bigquery/src/it/scala/com/spotify/scio/bigquery/BigQueryPartitionUtilIT.scala [hunk 6]
```diff
@@ -0,0 +1,108 @@
+package com.spotify.scio.bigquery
```
Comment (ravwojdyla) [inline scio-bigquery/src/it/scala/com/spotify/scio/bigquery/BigQueryPartitionUtilIT.scala]: Header missing.

**Comment by keithmcneill** [comment]
This will be great! Thanks!

**Pull request merged**